analysis/quantifier-comparisons/13-spatial-transcriptomics-benchmarking.Rmd

```diff
@@ -0,0 +1,411 @@
+
+```{r}
+# create one list with both spe's together
+all_spe_list <- list(fry_spe, spaceranger_spe)
+names(all_spe_list) <- c("alevin-fry", "spaceranger")
+
+# calculate per cell QC and output to a combined data frame with plotting 
+all_spe_list <- all_spe_list %>%
+    purrr::map(
+      ~ scuttle::addPerCellQCMetrics(.x, 
+                                     subsets = list(mito = mito_genes[mito_genes %in% rownames(.x)]))) %>%
+  purrr::map(scuttle::addPerFeatureQCMetrics)
+```
+
+After adding in the per spot QC metrics to both of the spe's, we want to extract the `colData` from each spe and create a data frame that we can use for plotting. 
+We will also need some information about each sample and how it was run, so we will create a sample metadata table, `sample_info_df` that will then be merged with the `colData`. 
+
+```{r}
+# create sample info dataframe to be joined with per spot dataframe later
+sample_info_df <- quant_info_table(data_dir= s3_out, 
+                 tools = c("cellranger", "alevin-fry-knee"),
+                 samples = sample_id) %>%
+  # convert cellranger to spaceranger 
+  dplyr::mutate(tool = ifelse(tool == "cellranger", "spaceranger", tool))
+```
+
+When we convert the `colData` to a data frame we use the custom function, `spatial_coldata_to_df()` to do so and apply it to each spe in our list. 
+```{r}
+# join coldata dataframe with sample info
+coldata_df <- purrr::map_df(all_spe_list, spatial_coldata_to_df, .id = "tool") %>%
+  # remove extra -1 from spaceranger barcodes
+  dplyr::mutate(spot_id = gsub("-1", "", spot_id)) %>%
+  dplyr::left_join(sample_info_df,
+                   by = c("tool"))
+```
+
+Now we only want to filter our data frame to contain spots that are shared between both tools and those that are found to be overlapping with the tissue. 
+
+```{r}
+# identify shared spots only 
+spot_counts <- coldata_df %>%  
+  dplyr::count(spot_id, sample)
+
+common_spots <- spot_counts %>%
+  dplyr::filter(n == 2) %>%
+  dplyr::pull(spot_id)
+
+coldata_df_common <- coldata_df %>%
+  dplyr::filter(spot_id %in% common_spots,
+                # only include spots that overlap with tissue
+                in_tissue == 1)
+```
+
+We will also need to filter the spe's directly based on spots that are present in the tissue, so we create a small function to do this and then apply it to both spe's in the list. 
+
+```{r}
+# we will also want to filter the spe's directly 
+filter_spe <- function(spe){
+  spe <- spe[, spatialData(spe)$in_tissue == 1]
+}
+
+all_spe_filter <- all_spe_list %>%
+  purrr::map(filter_spe)
+```
+
+When we look at our results, we will also want to visualize them so we will make a custom function to plot the results.
+Because of the bug that is present in `SpatialExperiment::read10XVisium()` function, we will make two separate functions, one for spaceranger only spes and one for alevin-fry + spaceranger spes.
```

Comment:
As I said earlier, I would rather work around this bug directly at read time (or avoid the function entirely).